apps/juxtaposition-ui/src/types/middleware.ts

```diff
@@ -0,0 +1,21 @@
+import type { Request } from 'express';
+import type { GetUserDataResponse as ApiGetUserDataResponse } from '@pretendonetwork/grpc/api/get_user_data_rpc';
+import type { GetUserDataResponse as AccountGetUserDataResponse } from '@pretendonetwork/grpc/account/get_user_data_rpc';
+import type { ParamPack } from '@/types/common/param-pack';
+
+export type AuthRequest<TReq extends Request = Request> = TReq & {
+	user: ApiGetUserDataResponse | AccountGetUserDataResponse;
+	pid: number;
+	tokens: {
+		serviceToken?: string;
+		oauthToken?: string;
+	};
+	paramPackData: null | ParamPack;
+};
+
+export function getAuthedRequest<TReq extends Request = Request>(req: TReq): AuthRequest<TReq> {
+	if (!(req as any).user) {
```

Comment:
I think this boxes us in a little, since if the middleware were to be converted next, it wouldn't be able to set `req.user`, since that's not in the base Request type.

I think it would be nicer to extend the base Request type so our custom bits are available everywhere. We already do this with the `directory` field: https://github.com/PretendoNetwork/juxtaposition/blob/dev/apps/juxtaposition-ui/src/types/express.d.ts#L6-L8
Also miiverse-api, where it's used for the same data: https://github.com/PretendoNetwork/juxtaposition/blob/dev/apps/miiverse-api/src/types/express.d.ts

Eventually it would be good to introduce something like the AccountData struct from `internal` everywhere (would simplify guest access a lot), but I recognise that's out of scope for this PR.

Reply:
We discussed this in DM's and concluded that we can improve it later. I added a TODO with a little detail.